File: codes/javascript/gpt/partial/task_4/Solution.js
function allocateResources(patients, resources, responders, totalTime) {
    // Step 1: Initialize variables
    let assignments = [];
    let availability = Array(responders).fill(0); // Track responder availability
    let total_survival_probability = 0;
    let survivors = 0;

    // Step 2: Helper function to calculate final probability after treatment
    function calculateFinalProbability(patient, startTime, resource) {
        let initialProb = patient[startTime] || 0;
        let finalProb = initialProb + resource.boost;
        return Math.min(finalProb, 100); // Probabilities cannot exceed 100%
    }

    // Step 3: Allocate resources
    for (let time = 0; time < totalTime; time++) {
        for (let patientIndex = 0; patientIndex < patients.length; patientIndex++) {
            let bestResource = null;
            let bestProbIncrease = 0;
            let startTime = -1;

            // Find the best resource for the current patient at the current time
            for (let resourceIndex = 0; resourceIndex < resources.length; resourceIndex++) {
                const resource = resources[resourceIndex];
                const treatmentTime = time + resource.time;

                if (treatmentTime <= totalTime && availability.includes(time)) {
                    const finalProb = calculateFinalProbability(patients[patientIndex], time, resource);
                    const probIncrease = finalProb - patients[patientIndex][time];

                    // Choose the resource that provides the highest probability increase
                    if (probIncrease > bestProbIncrease) {
                        bestProbIncrease = probIncrease;
                        bestResource = resourceIndex;
                        startTime = time;
                    }
                }
            }

            // Step 4: Assign the best resource if available
            if (bestResource !== null) {
                const responderIndex = availability.indexOf(time);
                availability[responderIndex] = time + resources[bestResource].time; // Update responder availability
                assignments.push({ patient: patientIndex, resource: bestResource, time: startTime });

                const finalProb = calculateFinalProbability(patients[patientIndex], startTime, resources[bestResource]);
                patients[patientIndex][startTime] = finalProb;

                // Update total survival probability
                total_survival_probability += finalProb;
            }
        }
    }

    // Step 5: Calculate survivors
    for (let patient of patients) {
        const finalProb = Math.max(...patient);
        if (finalProb > 50) {
            survivors++;
        }
    }

    // Step 6: Return the result
    return {
        assignments,
        survivors,
        total_survival_probability
    };
}

// Reading input from stdin
const fs = require('fs');
const input = JSON.parse(fs.readFileSync(0, 'utf8'));
const result = allocateResources(input.patients, input.resources, input.responders, input.total_time);

// Print result as JSON
console.log(JSON.stringify(result));

File: codes/go/gpt/partial/task_4/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

func allocateResources(patients [][]int, resources []map[string]interface{}, responders int, totalTime int) map[string]interface{} {
	assignments := []map[string]int{}
	survivors := 0
	totalSurvivalProbability := 0

	// Simulate a simple greedy allocation strategy
	availableResponders := make([]int, totalTime)
	resourceUsage := make([]bool, len(resources))
	for t := 0; t < totalTime; t++ {
		availableResponders[t] = responders
	}

	for time := 0; time < totalTime; time++ {
		for i, patient := range patients {
			if time < len(patient) && availableResponders[time] > 0 {
				bestResourceIdx := -1
				bestProbability := 0
				for j, resource := range resources {
					if resourceUsage[j] {
						continue
					}
					resourceTime := int(resource["time"].(float64))
					boost := int(resource["boost"].(float64))
					if time+resourceTime < len(patient) {
						newProbability := patient[time+resourceTime] + boost
						if newProbability > 100 {
							newProbability = 100
						}
						if newProbability > bestProbability {
							bestProbability = newProbability
							bestResourceIdx = j
						}
					}
				}
				if bestResourceIdx != -1 {
					availableResponders[time]--
					resourceUsage[bestResourceIdx] = true
					assignments = append(assignments, map[string]int{"patient": i, "resource": bestResourceIdx, "time": time})
					if bestProbability > 50 {
						survivors++
					}
					totalSurvivalProbability += bestProbability
				}
			}
		}
	}

	return map[string]interface{}{
		"assignments":             assignments,
		"survivors":               survivors,
		"total_survival_probability": totalSurvivalProbability,
	}
}

func main() {
	var input map[string]interface{}
	json.NewDecoder(os.Stdin).Decode(&input)

	patients := input["patients"].([]interface{})
	convertedPatients := make([][]int, len(patients))
	for i, p := range patients {
		patient := p.([]interface{})
		convertedPatients[i] = make([]int, len(patient))
		for j, prob := range patient {
			convertedPatients[i][j] = int(prob.(float64))
		}
	}

	resources := input["resources"].([]interface{})
	convertedResources := make([]map[string]interface{}, len(resources))
	for i, r := range resources {
		convertedResources[i] = r.(map[string]interface{})
	}

	responders := int(input["responders"].(float64))
	totalTime := int(input["total_time"].(float64))

	result := allocateResources(convertedPatients, convertedResources, responders, totalTime)
	output, _ := json.Marshal(result)
	fmt.Println(string(output))
}